dbms/src/Storages/DeltaMerge/ReadThread/ColumnSharingCache.h

```diff
@@ -0,0 +1,199 @@
+#pragma once
+
+#include <Columns/IColumn.h>
+#include <DataTypes/IDataType.h>
+#include <Storages/DeltaMerge/File/DMFile.h>
+
+#include <memory>
+
+namespace DB::DM
+{
+enum class ColumnCacheStatus
+{
+    ADD_COUNT = 0,
+    ADD_STALE,
+    GET_MISS,
+    GET_PART,
+    GET_HIT,
+    GET_COPY,
+
+    _TOTAL_COUNT,
+};
+
+class ColumnSharingCache
+{
+public:
+    struct ColumnData
+    {
+        size_t pack_count{0};
+        ColumnPtr col_data;
+    };
+
+    void add(size_t start_pack_id, size_t pack_count, ColumnPtr & col_data)
+    {
+        std::lock_guard lock(mtx);
+        auto & value = packs[start_pack_id];
+        if (value.pack_count < pack_count)
+        {
+            value.pack_count = pack_count;
+            value.col_data = col_data;
+        }
+    }
+
+    ColumnCacheStatus get(size_t start_pack_id, size_t pack_count, size_t read_rows, ColumnPtr & col_data, DataTypePtr data_type)
+    {
+        ColumnCacheStatus status;
+        std::lock_guard lock(mtx);
+        auto target = packs.find(start_pack_id);
+        if (target == packs.end())
+        {
+            status = ColumnCacheStatus::GET_MISS;
+        }
+        else if (target->second.pack_count < pack_count)
+        {
+            status = ColumnCacheStatus::GET_PART;
+        }
+        else if (target->second.pack_count == pack_count)
+        {
+            status = ColumnCacheStatus::GET_HIT;
+            col_data = target->second.col_data;
+        }
+        else
+        {
+            status = ColumnCacheStatus::GET_COPY;
+            auto column = data_type->createColumn();
+            column->insertRangeFrom(*(target->second.col_data), 0, read_rows);
+            col_data = std::move(column);
+        }
+        return status;
+    }
+
+    void del(size_t upper_start_pack_id)
+    {
+        std::lock_guard lock(mtx);
+        for (auto itr = packs.begin(); itr != packs.end();)
+        {
+            if (itr->first + itr->second.pack_count <= upper_start_pack_id)
+            {
+                itr = packs.erase(itr);
+            }
+            else
+            {
+                break;
+            }
+        }
+    }
+
+private:
+    std::mutex mtx;
+    // start_pack_id -> <pack_count, col_data>
+    std::map<size_t, ColumnData> packs;
+};
+
+class ColumnSharingCacheMap
```

Comment:
Please add some documents to explain the strategy of this cache. Like why calling del by upper_start_pack_id after get.